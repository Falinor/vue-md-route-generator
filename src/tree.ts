export type Tree<T> = { value: T; children?: Tree<T>[] }

type ForEachFunction<T> = (node: T) => void

export function forEach<T = any>(tree: Tree<T>, fn: ForEachFunction<T>): void {
  const { value, children } = tree
  fn(value)

  if (!children) {
    return
  }

  children.forEach(child => {
    forEach(child, fn)
  })
}

type MapFunction<T, U> = (value: T, children?: Tree<T>[]) => U

export function map<T = any, U = any>(
  tree: Tree<T>,
  fn: MapFunction<T, U>
): Tree<U> {
  const { value, children } = tree

  if (!children) {
    return { value: fn(value) }
  }

  return {
    value: fn(value, children),
    children: children.map(child => map(child, fn))
  }
}

export function flatten<T>(tree: Tree<T>): T[] {
  function flattenRec<T>(acc: T[], node: Tree<T>): T[] {
    acc.push(node.value)
    if (node.children) {
      acc.concat(...node.children.map(child => flattenRec(acc, child)))
    }
    return acc
  }
  return flattenRec<T>([], tree)
}

export function flatMap<T = any, U = any>(
  tree: Tree<T>,
  fn: MapFunction<T, U>
): U[] {
  return flatten(map(tree, fn))
}

export type SortFunction<T> = (a: Tree<T>, b: Tree<T>) => number

export function sort<T = any>(tree: Tree<T>, fn: SortFunction<T>): Tree<T> {
  const { value, children } = tree

  if (!children) {
    return tree
  }

  return {
    value,
    children: [...children].map(child => sort(child, fn)).sort(fn)
  }
}

export type FilterFunction<T> = (value: T, children?: Tree<T>[]) => boolean

export function filter<T = any>(
  tree: Tree<T>,
  fn: FilterFunction<T>
): Tree<T> | null {
  if (!fn(tree.value)) {
    return null
  }

  if (!tree.children) {
    return tree
  }

  return {
    value: tree.value,
    children: tree.children
      .map(child => filter(child, fn))
      .filter(v => v !== null) as Tree<T>[]
  }
}
