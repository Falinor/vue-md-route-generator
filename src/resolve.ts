import parseFrontMatter from 'gray-matter'
import isEmpty from 'lodash.isempty'
import path from 'path'

import { map, Tree } from './tree'

export interface PageMeta {
  name: string
  specifier: string
  path: string
  pathSegments: string[]
  component: string | null
  routeMeta?: any
}

export interface File {
  path: string
  type: 'directory' | 'file'
}

export type FileTree = Tree<File>
export type PageMetaTree = Tree<PageMeta>

export function resolveRoutePaths(
  tree: FileTree,
  importPrefix: string,
  readFile: (path: string) => string
): Tree<PageMeta> {
  return map(tree, (file: File) => {
    const segments: string[] = file.path.split('/')
    const meta: PageMeta = {
      name: pathToName(segments),
      path: pathToRoute(segments, 0, false),
      pathSegments: toActualPath(segments),
      component: null,
      specifier: pathToSpecifier(segments)
    }

    if (file.type === 'file') {
      const content = readFile(file.path)
      const { data } = parseFrontMatter(content)
      if (!isEmpty(data)) {
        meta.routeMeta = data
      }
      meta.component = importPrefix + file.path
    }

    return meta
  })
}

function isOmittable(segment: string): boolean {
  return segment === 'index'
}

/**
 * - Remove `.md` from the last path
 * - Omit if the last segment is `index`
 * - Convert dynamic route to `:param` format
 */
function toActualPath(segments: string[]): string[] {
  const lastIndex = segments.length - 1
  const last = basename(segments[lastIndex])

  const segmentsWithoutLast = segments.slice(0, -1)
  return isOmittable(last)
    ? segmentsWithoutLast
    : segmentsWithoutLast.concat(last)
}

function pathToName(segments: string[]): string {
  const last = segments[segments.length - 1]
  return segments
    .slice(0, -1)
    .concat(basename(last))
    .join('-')
}

function pathToSpecifier(segments: string[]): string {
  const name = pathToName(segments)
  const replaced = name
    .replace(/(^|[^a-zA-Z])([a-zA-Z])/g, (_, a, b) => {
      return a + b.toUpperCase()
    })
    .replace(/[^a-zA-Z0-9]/g, '')

  return /^\d/.test(replaced) ? '_' + replaced : replaced
}

function pathToRoute(
  segments: string[],
  parentDepth: number,
  nested: boolean
): string {
  const prefix = nested || parentDepth > 0 ? '' : '/'
  return (
    prefix +
    toActualPath(segments)
      .slice(parentDepth)
      .join('/')
  )
}

function basename(filename: string): string {
  return path.basename(filename, '.md')
}
