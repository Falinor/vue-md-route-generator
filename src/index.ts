import fs from 'fs'
import directoryTree, { DirectoryTree } from 'directory-tree'
import { basename, dirname, isAbsolute, join } from 'path'

import { createRoutes } from './template/routes'
import { resolveRoutePaths } from './resolve'
import { FileTree } from './resolve'

export interface GenerateConfig {
  folders: string[]
  importPrefix?: string
  dynamicImport?: boolean
  chunkNamePrefix?: string
  nested?: boolean
}

function relativeTo(prefix: string, path: string): string {
  const segments = path.split('/')
  const index = segments.findIndex(seg => seg === prefix)
  return segments.slice(index).join('/')
}

function toPathTree(directoryTree: DirectoryTree, prefix: string): FileTree {
  return {
    value: {
      path: isAbsolute(directoryTree.path)
        ? relativeTo(prefix, directoryTree.path)
        : directoryTree.path,
      type: directoryTree.type
    },
    children: directoryTree?.children?.map(child => toPathTree(child, prefix))
  }
}

export function generateRoutes({
  folders,
  importPrefix = '@/assets/',
  dynamicImport = true,
  chunkNamePrefix = ''
}: GenerateConfig): string {
  return folders
    .map(folder => {
      const dirTree = directoryTree(folder, { extensions: /\.md$/ })
      const pathTree = toPathTree(dirTree, basename(folder))

      const meta = resolveRoutePaths(pathTree, importPrefix, file => {
        return fs.readFileSync(join(dirname(folder), file), 'utf8')
      })
      return createRoutes(meta, dynamicImport, chunkNamePrefix)
    })
    .join(',\n')
}
