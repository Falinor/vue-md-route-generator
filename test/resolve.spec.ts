import { FileTree, resolveRoutePaths } from '../src/resolve'

function mockReadFile(path: string): string {
  if (path === 'meta.md') {
    return `---
title: Hello
---`
  }

  if (path === 'invalid-meta.vue') {
    return `---
{ invalid: Test }
---`
  }

  return ''
}

describe('Route resolution', () => {
  function test(name: string, tree: FileTree): void {
    it(name, () => {
      expect(
        resolveRoutePaths(tree, '@/assets/', mockReadFile)
      ).toMatchSnapshot()
    })
  }

  test('resolves routes', {
    value: { path: 'guides', type: 'directory' },
    children: [
      {
        value: { path: 'guides/foo.md', type: 'file' }
      },
      {
        value: { path: 'guides/bar.md', type: 'file' }
      },
      {
        value: { path: 'guides/baz.md', type: 'file' }
      }
    ]
  })

  test('resolves route meta', {
    value: {
      path: 'meta.md',
      type: 'file'
    }
  })
})
