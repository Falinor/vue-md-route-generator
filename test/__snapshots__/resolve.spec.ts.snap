// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Route resolution resolves route meta 1`] = `
Object {
  "value": Object {
    "component": "@/assets/meta.md",
    "name": "meta",
    "path": "/meta",
    "pathSegments": Array [
      "meta",
    ],
    "routeMeta": Object {
      "title": "Hello",
    },
    "specifier": "Meta",
  },
}
`;

exports[`Route resolution resolves routes 1`] = `
Object {
  "children": Array [
    Object {
      "value": Object {
        "component": "@/assets/guides/foo.md",
        "name": "guides-foo",
        "path": "/guides/foo",
        "pathSegments": Array [
          "guides",
          "foo",
        ],
        "specifier": "GuidesFoo",
      },
    },
    Object {
      "value": Object {
        "component": "@/assets/guides/bar.md",
        "name": "guides-bar",
        "path": "/guides/bar",
        "pathSegments": Array [
          "guides",
          "bar",
        ],
        "specifier": "GuidesBar",
      },
    },
    Object {
      "value": Object {
        "component": "@/assets/guides/baz.md",
        "name": "guides-baz",
        "path": "/guides/baz",
        "pathSegments": Array [
          "guides",
          "baz",
        ],
        "specifier": "GuidesBaz",
      },
    },
  ],
  "value": Object {
    "component": null,
    "name": "guides",
    "path": "/guides",
    "pathSegments": Array [
      "guides",
    ],
    "specifier": "Guides",
  },
}
`;
